### DISCUSSION TITLE
Where to go from here?

### DISCUSSION THREAD

**Discussion by Dakkaron** [Dakkaron/Fairberry#5] [Ideas]
So, the first working prototype is out, there is a lot of interest, but there are some issues:

- To build one, a lot of different skills and tools are necessary (3D printer, SMD soldering, maybe blender to adjust for a different phone)
- Due to all the hand soldered parts, durability might be an issue
- The current electronics can't be ordered in finished state (due to all the hand soldering and the use of an Arduino)
- The current Microcontroller (Atmega32u4) is a bit oversized, rather expensive (~€6) and hard to find
- We currently are totally relying on the BB Q10 keyboard, which is a problem, since the phone is EOL and the supply could dry up any time. I don't have a solution for that. Maybe we can find a different phone that is still available as a source for replacement parts?

Did I miss anything?

I have a few ideas for possible solutions, but I'd really like your input, so don't take my ideas as fact.

- Probably the most important step would be to create a custom PCB. That would allow us to order the electronics all in one piece, already with the SMD components soldered on. That means, the only soldering necessary would be for the USB connector, which would be through hole soldering, which even a novice solderer can do.
- We could consider putting the USB connector directly onto the board, so that it could be soldered directly by the PCB factory. That would mean, so soldering necessary, but also would make repears much harder if the connector breaks.
- As the microcontroller we could use an ATMega16u2, which is cheaper and much more easily available. It should also be code-compatible to the Atmega32u4. I don't have a lot of experience with differen Microcontrollers. If you know another one with enough pins, built-in pullups, native USB, good power management/low power consumption, low price, good availability and (if possible) Arduino compatibility, let me know!
- The 3D print file could be rebuilt as a configurable OpenSCAD model. That would mean, to adjust to a different phone, you'd only need to enter the dimensions of the phone and it would automatically generate the case for a different phone.

That should be enough, that someone who wants the keyboard can just order the parts ready-made. Only for the electronics it might make sense to order a bigger amount of pieces to cut down the price (JLCPCB has a setup cost of around €10-15 and a minimum order quantity of 5, if I remember correctly).

**Comment by spARTacusFromCB** [comment]
That all makes sense to me. I guess there needs to be an understanding as to what sort of cost / effort to get to there, and who to do what to make it all happen?

In one of your posts somewhere, you mentioned about maybe also incorporating pass through USB charging and headphone jack. I think it might be akward to try to accommodate too many different possible combinations of different phones, and although the headphone jack would be great for my current devices, I am guessing that only the pass through USB would be most universal.

Awesome work.

**Comment by Dakkaron** [comment]
The cost shouldn't be too high, the labor might be. KiKAD and OpenSCAD are both freeware (open source ftw!). The biggest problem would be, that both need quite a bit of skill and knowledge to operate. Being a programmer, OpenSCAD isn't too hard for me, so I think I should be able to do that part.

I tried learning KiKAD before. I can make a schematic in KiCAD, but I am lost on footprints and layouts. I considered paying someone on Fivr to do it, but I don't have any experience there and I don't know how that would turn out. Also, my budget for hobby stuff is kinda stretched thin, at the moment.

For ordering PCBs I used JLCPCB before. They are cheap and really good. I did some preliminary checking what the boards would cost if made and assembled by them. These numbers are very rough estimates, since they can only give me a proper quote once I have the PCB design files.

They would have fixed costs (no matter how many pcs are ordered) of €10-15. Each piece would then cost ~€10-15. Depending especially on how big the final PCB gets. As far as I remember, the minimum order quantity is 5, and I think you can only order in increments of 5 as well.

That would be only the electronics, without the keyboard and without the 3D print. Also, the shipment would only be delivered to a single address, and I'd need to distribute/ship to whereever the people live who are interested in it.

The keyboards themselves can be found on AliExpress, and since they ship from China, it is much cheaper if everyone orders them separately (shipping from China is going to be much cheaper than from Austria).

Regarding the 3D prints: I think it would be best if everyone finds a local place to print them, since they are pretty wide-spread by now, and often you can even use them for free or very cheap (e.g. in libraries or makerspaces). That way, you can also fine-tune and experiment with tolerances.

**Reply by spARTacusFromCB** [reply]
Any idea what these guys did:

https://www.tindie.com/products/lilygo/lilygo-ttgo-t-watch-keyboard/

**Reply by Dakkaron** [reply]
Sorry for the late answer. From what I can see, they are doing something similar. Only, their board doesn't output USB, but instead it looks like I²C. I²C totally makes sense, since it has far less overhead than USB. But I don't really know any Android phone that has an I2C connection...

Another way to go would be to use this here: https://github.com/arturo182/pmod_bbq10_keyboard

And then connect that to some Arduino which then converts I²C to USB. But that takes more space and uses even more power, since now it has to power two separate microcontrollers.

I am currently considering to create a schematic and pay someone on Fivr to then design the PCB.

It shouldn't be too hard, to take the board from my link above, replace the microcontroller with something that is able to output USB and change the output pins from I²C to USB.

**Comment by spARTacusFromCB** [comment]
What about what these guys did? Looks like they may have even tried to do something for a trackpad.

https://www.tindie.com/products/electronictrik/blackberry-q10-kb-prototyping-breadboard/

**Reply by Dakkaron** [reply]
This is just a breakout board, that also contains an OLED. It's very similar to the breakout board that is used in the Fairberry project.

**Comment by spARTacusFromCB** [comment]
Hmmm. This looked interesting also.

https://www.cartft.com/catalog/il/478

**Reply by Dakkaron** [reply]
This looks very interesting. Looks to be out of production though, from what I can see. But it could work for one-offs. I think I'll order one.

**Reply by spARTacusFromCB** [reply]
I did a bunch of searching about it earlier, and the guy apparently had maybe hundreds of them back in 2017. I don't have confidence as to how many might still be around/available, or what that site is about for ordering from. However, I think at least one post indicated stock available (but maybe that was an old status).

**Comment by spARTacusFromCB** [comment]
What are your thoughts about the alternate PCB you are contemplating, in terms of the USB-C connection/connector? Would the USB-C connector still have a bit of a cable lead for it or would it be hard soldered in place?

For your currently prototype, your keyboard sits higher than the surface of the phone and the USB-C I think connects underneath. In my scenario, I'd be planning to modify an existing case and have the keyboard surface almost flush with the phone surface. However, that might get awkward depending on how/where the USB-C connector sits if it is hard mounted on a board, as opposed to still with a bit of a wire lead.

Presumably, the placement of the USB-C ports on different Android slabs or tablets (not sure if all of them are exactly centred horizontally and depth wise) might mean a hard soldered USB-C connector could impact compatibility for the hardware for maximum possible of different devices. At minimum, presumably the depth placement of the USB-C port on different devices are all at least slightly different by maybe a few mm, even if all of them are horizontally centred.

Any thoughts?

**Reply by Dakkaron** [reply]
Tbh, I don't like the hard soldered connector. It will fail faster and be much less reparable. If you manage to strip the leads, where the connector is soldered to, from the PCB, you can basically toss the PCB.

There are only two advantages to the hard-soldered connector.
1) They can be assembled by the PCB factory
2) You don't neet to manually connect the connector to the phone, since attaching the attachment to the phone will automatically connect the connector.


But

1) doesn't matter too much, since soldering the USB cable is the easiest solder job on the whole device anyways.
2) can also be accomplished by hard-mounting the USB-C connector to the case of the attachment.

Other question: How good are your soldering skills? If you want to, I can try to assemble one of the adapters and send that to you. But I don't have a spare Arduino Pro Micro, so you'd have to solder the connections between adapter and Arduino, and also Arduino to USB yourself. Are you interested?

**Reply by spARTacusFromCB** [reply]
I really don't have much soldering ability. I can resolder a speaker wire and stuff like that for some minor repair work, but that is about it. I have never soldered components and the soldering iron I have is probably like a baseball bat compared to what is normally used for small components and precise work.

I am not sure it would be a good idea to send me anything right now. Maybe we keep that in mind for later, if we have to.

I have tried to drum up more interest on CB, and for identifying those that also have more technical capabilities and skills to help. However, it has not yet produced a lot.

**Reply by spARTacusFromCB** [reply]
Hole soldering a USB-C connector (with lead wires) to a board would probably be within my stretched soldering ability.

**Comment by dumblob** [comment]
We have now a new "competitor" product :wink:.

https://www.clicks.tech/

It looks really good to me. Unfortunately they offer only iPhone keyboards for now. Maybe later their portfolio will increase, who knows (if they will not go bankrupt in the mean time as the market is probably pretty small).
